core/src/main/java/com/graphhopper/reader/osm/OSMReader.java

```diff
@@ -576,53 +576,49 @@ static List<OSMTurnRelation> createTurnRelations(ReaderRelation relation) {
+            osmTurnRelations.addAll(createTurnRelations(relation, relation.getTag("restriction"),
+                    vehicleTypeRestricted, vehicleTypesExcept));
             return osmTurnRelations;
         }
         if (relation.hasTagWithKeyPrefix("restriction:")) {
             List<String> vehicleTypesRestricted = relation.getKeysWithPrefix("restriction:");
             for (String vehicleType : vehicleTypesRestricted) {
                 String restrictionType = relation.getTag(vehicleType);
                 vehicleTypeRestricted = vehicleType.replace("restriction:", "").trim();
-                OSMTurnRelation osmTurnRelation = createTurnRelation(relation, restrictionType, vehicleTypeRestricted, vehicleTypesExcept);
-                if (osmTurnRelation != null) {
-                    osmTurnRelations.add(osmTurnRelation);
-                }
+                osmTurnRelations.addAll(createTurnRelations(relation, restrictionType, vehicleTypeRestricted,
+                        vehicleTypesExcept));
             }
         }
         return osmTurnRelations;
     }
 
-    static OSMTurnRelation createTurnRelation(ReaderRelation relation, String restrictionType, String
-            vehicleTypeRestricted, List<String> vehicleTypesExcept) {
+    static List<OSMTurnRelation> createTurnRelations(ReaderRelation relation, String restrictionType,
+                                                     String vehicleTypeRestricted, List<String> vehicleTypesExcept) {
         OSMTurnRelation.Type type = OSMTurnRelation.Type.getRestrictionType(restrictionType);
         if (type != OSMTurnRelation.Type.UNSUPPORTED) {
-            long fromWayID = -1;
             long viaNodeID = -1;
             long toWayID = -1;
 
             for (ReaderRelation.Member member : relation.getMembers()) {
-                if (ReaderElement.WAY == member.getType()) {
-                    if ("from".equals(member.getRole())) {
-                        fromWayID = member.getRef();
-                    } else if ("to".equals(member.getRole())) {
-                        toWayID = member.getRef();
-                    }
-                } else if (ReaderElement.NODE == member.getType() && "via".equals(member.getRole())) {
+                if (ReaderElement.WAY == member.getType() && "to".equals(member.getRole()))
+                    toWayID = member.getRef();
+                else if (ReaderElement.NODE == member.getType() && "via".equals(member.getRole()))
                     viaNodeID = member.getRef();
-                }
             }
-            if (fromWayID >= 0 && toWayID >= 0 && viaNodeID >= 0) {
-                OSMTurnRelation osmTurnRelation = new OSMTurnRelation(fromWayID, viaNodeID, toWayID, type);
-                osmTurnRelation.setVehicleTypeRestricted(vehicleTypeRestricted);
-                osmTurnRelation.setVehicleTypesExcept(vehicleTypesExcept);
-                return osmTurnRelation;
+            if (toWayID >= 0 && viaNodeID >= 0) {
```

Comment:
Not really related to this PR probably, but we just had this discussion in the forum: https://discuss.graphhopper.com/t/custom-turn-restriction-example/7453

Why do we check if `toWayID >= 0 && viaNodeID >= 0` and why do we expect negative OSM IDs here?

Reply:
Ah, I see now... But this means if an OSM ID is actually negative the turn relation will be ignored without any warning or error (see forum discussion). It's like a bad way to check if there was a relation member with the from/via/to role.

Reply:
Ok, then let's say it is not as it should be. I called it 'bad', because it only works when a certain assumption is fulfilled (all OSM IDs are >=0), and we have no way to control this assumption (because the IDs come from an external file), we do not check if our assumption is true, and there is not even a comment in the code that states that it only works under this assumption. It is just a bug waiting to happen, hidden in the code, and my guess is that when the code was written this case was just forgotten about. I would also use `null` here, and additionally I think we should throw an error early if an OSM ID is negative, or are negative OSM IDs a thing? It does not seem to be:

The OSM [wiki](https://wiki.openstreetmap.org/wiki/Node) says that the id is even >= 1 and

> Editors may temporarily save node ids as negative to denote ids that haven't yet been saved to the server.